pkg/operators/klusterlet/controllers/bootstrapcontroller/bootstrapcontroller.go

```diff
@@ -0,0 +1,176 @@
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	coreinformer "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	corelister "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/klog/v2"
+)
+
+// bootstrapController watches bootstrap hub kubeconfig secret, if the secret is changed with hub kube-apiserver ca or apiserver
+// endpoints, this controller will make the klusterlet re-bootstrap to get the new hub kubeconfig from hub cluster by deleting
+// the current hub kubeconfig secret and restart the klusterlet agents
+type bootstrapController struct {
+	kubeClient       kubernetes.Interface
+	klusterletLister operatorlister.KlusterletLister
+	secretLister     corelister.SecretLister
+}
+
+// NewBootstrapController returns a bootstrapController
+func NewBootstrapController(
+	kubeClient kubernetes.Interface,
+	klusterletInformer operatorinformer.KlusterletInformer,
+	secretInformer coreinformer.SecretInformer,
+	recorder events.Recorder) factory.Controller {
+	controller := &bootstrapController{
+		kubeClient:       kubeClient,
+		klusterletLister: klusterletInformer.Lister(),
+		secretLister:     secretInformer.Lister(),
+	}
+	return factory.New().WithSync(controller.sync).
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, err := meta.Accessor(obj)
+			if err != nil {
+				return ""
+			}
+			if accessor.GetName() == helpers.BootstrapHubKubeConfig {
+				return accessor.GetName()
+			}
+			return ""
+		}, secretInformer.Informer()).
+		ToController("BootstrapController", recorder)
+}
+
+func (k *bootstrapController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	bootstrapSecretName := controllerContext.QueueKey()
+	if bootstrapSecretName == "" {
+		return nil
+	}
+
+	klog.Infof("Reconciling bootstrap hub kubeconfig secret %q", bootstrapSecretName)
+
+	klusterlets, err := k.klusterletLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	errs := []error{}
+	for _, klusterlet := range klusterlets {
+		if err := k.reloadKlusterlet(ctx, controllerContext, klusterlet); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return nil
+}
+
+// reloadKlusterlet reload klusterlet by removing the current hub kubeconfig secret and klusterlet agent deployments
+func (k *bootstrapController) reloadKlusterlet(
+	ctx context.Context,
+	controllerContext factory.SyncContext,
+	klusterlet *operatorapiv1.Klusterlet) error {
+	namespace := klusterlet.Spec.Namespace
+	bootstrapHubKubeconfigSecret, err := k.secretLister.Secrets(namespace).Get(helpers.BootstrapHubKubeConfig)
```

Comment:
i would get namespace from key and compare with klusterlet namespace, we do not need to get secret until we know there is a klusterlet deployed in this namespace.

Reply:
done